waspc/src/Wasp/SemanticVersion.hs

```diff
@@ -78,27 +106,74 @@ instance Semigroup Range where
 instance Monoid Range where
   mempty = Range []
 
+isVersionInRange :: Version -> Range -> Bool
+isVersionInRange version (Range compSets) =
+  any (doesVersionSatisfyComparatorSet version) compSets
+
+doesVersionSatisfyComparatorSet :: Version -> ComparatorSet -> Bool
+doesVersionSatisfyComparatorSet version comparatorSet =
+  doesVersionSatisfyPrimitiveComparatorSet version $
+    comparatorSetToPrimitiveComparatorSet comparatorSet
+
+doesVersionSatisfyPrimitiveComparatorSet :: Version -> PrimitiveComparatorSet -> Bool
+doesVersionSatisfyPrimitiveComparatorSet version (PrimitiveComparatorSet comps) =
+  all (doesVersionSatisfyPrimitiveComparator version) comps
+
+doesVersionSatisfyPrimitiveComparator :: Version -> PrimitiveComparator -> Bool
+doesVersionSatisfyPrimitiveComparator version (PrimitiveComparator operator compVersion) =
+  case operator of
+    Equal -> version == compVersion
+    LessThan -> version < compVersion
+    LessThanOrEqual -> version <= compVersion
+    GreaterThan -> version > compVersion
+    GreaterThanOrEqual -> version >= compVersion
+
+comparatorSetToPrimitiveComparatorSet :: ComparatorSet -> PrimitiveComparatorSet
+comparatorSetToPrimitiveComparatorSet (ComparatorSet comps) =
+  foldl1 (<>) $ comparatorToPrimitiveComparatorSet <$> comps
+
+comparatorToPrimitiveComparatorSet :: Comparator -> PrimitiveComparatorSet
+comparatorToPrimitiveComparatorSet (ComparatorP comp) =
+  PrimitiveComparatorSet $ fromList [comp]
+comparatorToPrimitiveComparatorSet (ComparatorS comp) =
+  specialComparatorToPrimitiveComparatorSet comp
+
+specialComparatorToPrimitiveComparatorSet :: SpecialComparator -> PrimitiveComparatorSet
+specialComparatorToPrimitiveComparatorSet (Caret version) =
+  PrimitiveComparatorSet $
+    NE.fromList
+      [ PrimitiveComparator GreaterThanOrEqual version,
+        PrimitiveComparator LessThan (nextBreakingChangeVersion version)
+      ]
+
 nextBreakingChangeVersion :: Version -> Version
 nextBreakingChangeVersion version = case version of
   (Version 0 0 x) -> Version 0 0 (succ x)
   (Version 0 x _) -> Version 0 (succ x) 0
   (Version x _ _) -> Version (succ x) 0 0
 
-doesVersionSatisfyComparator :: Version -> Comparator -> Bool
-doesVersionSatisfyComparator version (Comparator operator compVersion) = case operator of
-  Equal -> version == compVersion
-  LessThanOrEqual -> version <= compVersion
-  BackwardsCompatibleWith -> version >= compVersion && version < nextBreakingChangeVersion compVersion
+-- Helper methods.
 
-doesVersionSatisfyComparatorSet :: Version -> ComparatorSet -> Bool
-doesVersionSatisfyComparatorSet version (ComparatorSet comps) = all (doesVersionSatisfyComparator version) comps
+(&) :: ComparatorSet -> ComparatorSet -> ComparatorSet
+(&) = (<>)
```

Comment:
Looks like we're not using this thing anywhere. And, even if we were, I'm sure if I'd vote to keep it.

`&` has a pretty standard semantic meaning (and, binary and, set intersection), I think our use case is too specific to adopt it (although I can see the allure). I'm aware other Haskell code uses it for its purposes, but I'm not a fan of it there either :).

If we want a name that's more specific than `<>` (not sure we do), can we use `union`? Oh, and that too. We have comparator **sets**, but using an `and/intersect` operation on them internally acts as a union (assuming my other comment holds). Seems a bit strange.

Reply:
Makes sense, removed!